File: Difficulty: Hard/LRU Cache/lru-cache.js
//{ Driver Code Starts
//Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let capacity = parseInt(readLine());
        let cache = new LRUCache(capacity);
        let queries = parseInt(readLine());
        let input_ar1 = readLine().split(' ');
        let index = 0;
        let s = '';
        while(queries--){
            let q = input_ar1[index++];
            if(q === "SET"){
                let key = parseInt(input_ar1[index++]);
                let value = parseInt(input_ar1[index++]);
                cache.set(key, value);
            }
            else{
                let key = parseInt(input_ar1[index++]);
                s += cache.get(key) + " ";
            }
        }
        console.log(s);
        
    }
}
// } Driver Code Ends


class LRUCache {
    constructor(cap) {
        this.capacity = cap;
        this.cache = new Map();
    }

    get(key) {
        if (!this.cache.has(key)) return -1;
        const value = this.cache.get(key);
        this.cache.delete(key);
        this.cache.set(key, value);
        return value;
    }

    set(key, value) {
        if (this.cache.has(key)) this.cache.delete(key);
        this.cache.set(key, value);
        if (this.cache.size > this.capacity) {
            const leastUsedKey = this.cache.keys().next().value;
            this.cache.delete(leastUsedKey);
        }
    }
}

File: Difficulty: Hard/LRU Cache/lru-cache.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// design the class in the most optimal way

class LRUCache {
  private:
  public:
  
  int lim = 0;
  int ctr=0;
  unordered_map<int,int> mp;
  unordered_map<int,int> del;
    // Constructor for initializing the cache capacity with the given value.
    LRUCache(int cap) {
        // code here
        lim = cap;
    }

    // Function to return value corresponding to the key.
    int get(int key) {
        // your code here
        if(mp.count(key))
        {
            del[key] = ctr++;
            return mp[key];
        }
        return -1;
    }

    // Function for storing key-value pair.
    void put(int key, int value) {
        // your code here
        if(mp.size() == lim)
        {
            if(mp.count(key))
            {
                mp[key] = value;
                del[key] = ctr;
            }
            else
            {
                int mini = INT_MAX;
                int k = -1;
                for(auto it : del)
                {
                    if(it.second < mini)
                    {
                        mini=it.second;
                        k= it.first;
                    }
                    // we got the key with minimumtime
                }
                if(k!=-1)
                {
                    mp.erase(k);
                    del.erase(k);
                }
                mp[key] = value;
                del[key] = ctr;
            }
        }
        else
        {
                mp[key] = value;
                del[key] = ctr;           
        }
        ctr++;
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {

        int capacity;
        cin >> capacity;
        LRUCache *cache = new LRUCache(capacity);

        int queries;
        cin >> queries;
        while (queries--) {
            string q;
            cin >> q;
            if (q == "PUT") {
                int key;
                cin >> key;
                int value;
                cin >> value;
                cache->put(key, value);
            } else {
                int key;
                cin >> key;
                cout << cache->get(key) << " ";
            }
        }
        cout << endl;
        cout << "~" << endl;
    }
    return 0;
}

// } Driver Code Ends